### PR TITLE
Propose project / working group for LLM Observability Semantic Conventions

### PR DESCRIPTION
**Pull request by drewby**
This proposes a project and working group for establishing Semantic Conventions for LLM applications. 

This project aims to establish observability standards for Large Language Model (LLM) applications, focusing on API interactions and multi-step LLM frameworks.

### CODE REVIEW

**Comment by sudivate** [comment]
@drewby I'm interested in this initiative please add me to the review group.

**Comment by drewby** [comment]
> @drewby I'm interested in this initiative please add me to the review group.

done.

File: projects/llm-semconv.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+**Maintainers:**
+
+- To be determined
+
+## Meeting Times
+
+Alternating weekly meetings to accommodate different time zones:
+
+- Week 1: Mondays at 8 AM PST
```
Comment (lmolkova) [inline projects/llm-semconv.md:51]: this conflicts with OTel SemConv meeting and it's unlikely I and other people involved into OTel semconv will be able to make this time work.  We can tentatively keep it and see how it works though.

File: projects/llm-semconv.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+- Week 2: Mondays at 4 PM PST
+
+## Timeline
+
+1. Initial Draft: February 2024
+2. Review and Refinement: April 2024
```
Comment (lmolkova) [inline projects/llm-semconv.md:57]: Can  we add prototypes somewhere between April/June?

I'm volunteering to help prototype Azure OpenAI SDKs in .NET and/or Java, it'd be great to have prototype in Python

**Comment by cartermp** [comment]
@drewby I'd like to be in the group as well. I have much more time to work on this now, although I don't think it's enough to be in a leader capacity. I ... think I can count as a sponsoring member as well? If so, please add me there.

File: projects/llm-semconv.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+## Description
+
+This project aims to establish observability standards for Large Language Model (LLM) applications, focusing on API interactions and multi-step LLM frameworks. Key areas include:
+
+- Developing semantic conventions for standard and complex LLM interactions.
+- Initiating vendor-specific conventions as a foundation for broader contributions.
+- Concentrating on attributes relevant to Metrics and Spans.
+
+For Spans: Instrumentation and debugging guidance for multi-step LLM applications, with recommendations on user privacy and intellectual property security, marking sensitive attributes and events as opt-in.
```
Comment (cartermp) [inline projects/llm-semconv.md:11]: I think this:

> marking sensitive attributes and events as opt-in

Will continue to be a point of discussion, both on the opt-in/out nature and where it's represented. I believe in my initial PR the consensus was that Logs were the best signal for holding critical debugging information like the full prompt and LLM response.

File: projects/llm-semconv.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
+# LLM Observability Semantic Conventions
+
+## Description
+
+This project aims to establish observability standards for Large Language Model (LLM) applications, focusing on API interactions and multi-step LLM frameworks. Key areas include:
```
Comment (cartermp) [inline projects/llm-semconv.md:5]: I think it would be good to also list some use cases for this information. Debugging is one, but so are evaluations (especially online evals), and so is acting as a data source for fine-tuning.

**Review by cartermp — COMMENTED** [review]
Some minor comments

**Comment by trask** [comment]
> I ... think I can count as a sponsoring member as well?

currently the sponsoring members need to be TC/GC or someone those groups have specifically agreed to delegate that responsibility to (in this case @lmolkova)

I believe @alolita wants to sponsor this group as well, which would give it the required 2 sponsors

File: projects/llm-semconv.md (hunk 2 above)
Comment (cartermp) [reply projects/llm-semconv.md:57]: For Python we could perhaps lean on OpenLLMetry? https://github.com/traceloop/openllmetry

I think it'll also be important to prototype something that uses OpenAI's Node SDK and/or event have a runtime-agnostic instrumentation package. JS is a quiet-but-big part of OpenAI usage from my standpoint.

File: projects/llm-semconv.md (hunk 1 above)
Comment (drewby) [reply projects/llm-semconv.md]: Updated the time for later in the week.

File: projects/llm-semconv.md (hunk 2 above)
Comment (drewby) [reply projects/llm-semconv.md]: Added this to the timeline

File: projects/llm-semconv.md (hunk 3 above)
Comment (drewby) [reply projects/llm-semconv.md:18]: Agreed. This is somewhere we'll need much more input and be willing to evolve as we learn more and prototype around the conventions. I like the Events/Logs approach, but I only know the scenarios I'm directly involved with.

File: projects/llm-semconv.md (hunk 4 above)
Comment (drewby) [reply projects/llm-semconv.md:5]: Great idea! I added some use cases. Let me know if they hit the mark.

**Comment by drewby** [comment]
> @drewby I'd like to be in the group as well. I have much more time to work on this now, although I don't think it's enough to be in a leader capacity. I ... think I can count as a sponsoring member as well? If so, please add me there.

Thanks, really glad you are going to be involved. I added you!

**Comment by drewby** [comment]
> I believe @alolita wants to sponsor this group as well, which would give it the required 2 sponsors

I confirmed with @alolita and added her to the list.

**Comment by alolita** [comment]
@trask @lmolkova @drewby Confirming my sponsorship as OTel GC member for the LLM Observability Semantic Conventions working group. Look forward to working with all of you! :-)

File: projects/llm-semconv.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+**Project Leads:**
+
+- @drewby
+- @nirga
+
+**Sponsoring Members:**
+
+- @lmolkova
+- @alolita
+- *Seeking additional sponsors*
```
Comment (lmolkova) [inline projects/llm-semconv.md:43]:
```suggestion
```

2 sponsors should be enough (unless someone wants to be added anyway)

File: projects/llm-semconv.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+**Maintainers:**
+
+- To be determined
+
+## Meeting Times
+
+Alternating weekly meetings to accommodate different time zones:
+
+- Week 1: Thursdays at 8 AM PST
```
Comment (lmolkova) [inline projects/llm-semconv.md:60]: I can't make Thu 8am work. 

Could we add this poll and try to find the most suitable time?

```suggestion
https://strawpoll.com/3RnYlN4Jxye
```

**Comment by drewby** [comment]
@alolita @sudivate @cartermp @nirga 

Can you add your preferred meeting times to this poll: https://strawpoll.com/3RnYlN4Jxye

**Comment by alolita** [comment]
Voted on the poll :-) @lmolkova @drewby

**Review by lmolkova — APPROVED** [review]
LGTM!

**Comment by lmolkova** [comment]
Created slack channel for the LLM WG: https://cloud-native.slack.com/archives/C06KR7ARS3X

**Review by cartermp — APPROVED** [review]

**Review by nirga — APPROVED** [review]

File: projects/llm-semconv.md
Comment (alolita) [inline projects/llm-semconv.md:1]: >>This project aims to establish observability standards for Large Language Model (LLM) applications, focusing on API interactions and multi-step LLM frameworks. 

I would reword this sentence to be clear about what we're aiming to cover. The updated sentences could be something like:

"This workgroup aims to identify and define a standardized vocabulary and specification for observability of Large Language Model (LLM) based applications. The effort would explore LLM based various application APIs, LLM tasks and LLM workflows for multistage reasoning."

Comment (alolita) [inline projects/llm-semconv.md:1]: >> - Developing semantic conventions for standard and complex LLM interactions
>> - Initiating vendor-specific conventions as a foundation for broader contributions
>> - Concentrating on attributes relevant to Metrics and Spans

I suggest changing to:
- Evaluating existing vendor-specific conventions (which may not be currently standardized but implementation specific)
- Developing a standardized vocabulary for observability specific semantic conventions for single stage and multistage LLM workflows
- Initial scope focused on metrics and traces (which could later be expanded to other telemetry data signals)

Comment (alolita) [inline projects/llm-semconv.md:1]: - Experimental semantic conventions for LLMs, allowing flexibility for evolution in this fast-paced field.
- Reference prototypes demonstrating the application of these conventions in data collection, processing, and analysis.
- Guidelines for monitoring LLM applications, including implementation strategies and privacy/security considerations.

I recommend 
* Experimental is more of a versioning process not necessarily the output of the workgroup. I would remove "experimental" from the deliverable. 
* Separate observability from security and privacy. Observability can support both security and privacy but would still have its own conventions.
* Instead of "guidelines" use "recommended practices".

**Review by alolita — COMMENTED** [review]
@drewby @lmolkova Have added some suggestions for review. Take a look.

File: projects/llm-semconv.md
Comment (drewby) [reply projects/llm-semconv.md:1]: Thank you! All great feedback and I've made the changes.

**Review by alolita — APPROVED** [review]
Lgtm - thanks for making the changes.

**Review by trask — APPROVED** [review]
